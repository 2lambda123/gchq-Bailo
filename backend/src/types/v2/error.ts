import Logger from 'bunyan'

export interface BailoError extends Error {
  // Inherited from 'Error'
  // name: string
  // message: string

  // An HTTP response code that represents the error
  code: number

  // This data is logged publicly to the frontend
  context?: {
    documentationUrl?: string

    [key: string]: unknown
  }

  // A custom logger may be provided, otherwise a default is used
  logger?: Logger
}
