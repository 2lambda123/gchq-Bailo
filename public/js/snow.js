const TOTAL_SNOWFLAKES = 200

let snowDiv = document.getElementById('snowContainer')

function embRand(a, b) {
  return Math.floor(Math.random() * (b - a + 1)) + a
}

if (!snowDiv) {
  const snowflakes = new Array(TOTAL_SNOWFLAKES).fill(`<i class="snow"></i>`).join('\n')

  const snow = `
    .snow {
      position: absolute;
      width: 10px;
      height: 10px;
      background: white;
      border-radius: 50%;
      margin-top:-10px;
    }

    #snowContainer {
      position: fixed;
      left: 0;
      top: 0;
      bottom: 0;
      width: 100vw;
      height: 100vh;
      overflow: hidden;
      z-index: 9999999;
      pointer-events: none;
    }
  `

  const transitions = new Array(TOTAL_SNOWFLAKES)
    .fill(undefined)
    .map((_, i) => {
      const rndX = embRand(0, 1000000) * 0.0001
      const rndO = embRand(-100000, 100000) * 0.0001
      const rndT = (embRand(3, 8) * 10).toFixed(2)
      const rndS = (embRand(0, 10000) * 0.0001).toFixed(2)

      return `
      .snow:nth-child(${i}) {
        opacity: ${(embRand(1, 10000) * 0.0001).toFixed(2)};
        transform: translate(${rndX.toFixed(2)}vw,-10px) scale(${rndS});
        animation: fall-${i} ${embRand(10, 30)}s -${embRand(0, 30)}s linear infinite;
      }

      @keyframes fall-${i}{
        ${rndT}%{
          transform: translate(${(rndX + rndO).toFixed(2)}vw,${rndT}vh) scale(${rndS});
        } to {
          transform: translate(${(rndX + rndO / 2).toFixed(2)}vw, 105vh) scale(${rndS});
        }
      }
    `
    })
    .join('\n')

  snowDiv = document.createElement('div')
  snowDiv.id = 'snowContainer'

  snowDiv.innerHTML = `
    <style>
      ${snow}
      ${transitions}
    </style>
    ${snowflakes}
  `
  document.body.appendChild(snowDiv)
}
