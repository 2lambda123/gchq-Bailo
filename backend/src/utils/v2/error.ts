import Logger from 'bunyan'

import { BailoError } from '../../types/v2/error.js'

export function GenericError(code: number, message: string, context?: BailoError['context'], logger?: Logger) {
  const err = Error(message) as BailoError

  if (err.stack) {
    // Remove this file from stack traces.
    err.stack = err.stack
      .split('\n')
      .filter((line) => !line.includes('backend/src/utils/v2/error.ts'))
      .join('\n')
  }

  err.code = code
  err.context = context

  Object.defineProperty(err, 'logger', {
    value: logger,
    enumerable: false,
  })

  return err
}

export function BadReq(message: string, context?: BailoError['context'], logger?: Logger) {
  return GenericError(400, message, context, logger)
}

export function Unauthorized(message: string, context?: BailoError['context'], logger?: Logger) {
  return GenericError(401, message, context, logger)
}

export function Forbidden(message: string, context?: BailoError['context'], logger?: Logger) {
  return GenericError(403, message, context, logger)
}

export function NotFound(message: string, context?: BailoError['context'], logger?: Logger) {
  return GenericError(404, message, context, logger)
}

export function InternalError(message: string, context?: BailoError['context'], logger?: Logger) {
  return GenericError(500, message, context, logger)
}

export function ConfigurationError(message: string, context?: BailoError['context'], logger?: Logger) {
  return GenericError(500, `BAILO configuration error: ${message}`, context, logger)
}
