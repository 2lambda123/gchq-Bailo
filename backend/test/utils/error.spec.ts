import { describe, expect, test } from 'vitest'

import { BadReq, Forbidden, GenericError, NotFound, Unauthorized } from '../../src/utils/v2/error.js'

describe('utils > error', () => {
  test('GenericError', () => {
    const error = GenericError(999, 'Example Message', { example: 'context' })

    expect(JSON.stringify(error, null, 4)).toMatchSnapshot()
    expect(error.message).toMatchSnapshot()
  })

  test('BadReq', () => {
    const error = BadReq('example')
    expect(error.code).toBe(400)
  })

  test('Unauthorized', () => {
    const error = Unauthorized('example')
    expect(error.code).toBe(401)
  })

  test('Forbidden', () => {
    const error = Forbidden('example')
    expect(error.code).toBe(403)
  })

  test('NotFound', () => {
    const error = NotFound('example')
    expect(error.code).toBe(404)
  })
})
