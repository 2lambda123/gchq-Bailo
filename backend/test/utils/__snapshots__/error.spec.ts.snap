// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`utils > error > GenericError 1`] = `
"{
    "code": 999,
    "context": {
        "example": "context"
    }
}"
`;

exports[`utils > error > GenericError 2`] = `"Example Message"`;
